### PR TITLE
Add traced datastore

### PR DESCRIPTION
**Pull request by dennis-tra**
This pull request introduces the package `trace` that wraps a datastore and applies open telemetry tracing to all datastore interactions.

### CODE REVIEW

File: trace/trace.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+		return dstore.Check(ctx)
+	}
+
+	return nil
+}
+
+// NewTransaction implements the ds.TxnDatastore interface.
+func (t *Datastore) NewTransaction(ctx context.Context, readOnly bool) (ds.Txn, error) {
+	ctx, span := t.tracer.Start(ctx, "NewTransaction", otel.WithAttributes(attribute.Bool("readOnly", readOnly)))
+	defer span.End()
```
Comment (Stebalien) [inline trace/trace.go:222]: Is there any way to associate gets/puts within the transaction with the transaction's span?

File: trace/trace.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	_ ds.CheckedDatastore    = (*Datastore)(nil)
+	_ ds.ScrubbedDatastore   = (*Datastore)(nil)
+	_ ds.GCDatastore         = (*Datastore)(nil)
+	_ io.Closer              = (*Datastore)(nil)
+)
+
+// Put implements the ds.Datastore interface.
+func (t *Datastore) Put(ctx context.Context, key ds.Key, value []byte) (err error) {
+	ctx, span := t.tracer.Start(ctx, "Put", otel.WithAttributes(attribute.String("key", key.String())))
+	defer span.End()
```
Comment (Stebalien) [inline trace/trace.go:42]: We're not tracing errors anywhere. We probably want to call `span.RecordError(err)` for all errors here (it'll ignore nil errors).

**Review by Stebalien — COMMENTED** [review]
Two nits, but this otherwise looks great!

File: trace/trace.go (hunk 2 above)
Comment (dennis-tra) [reply trace/trace.go:42]: ah good point, I guess it also makes sense to set the status then:

```
	// RecordError will record err as an exception span event for this span. An
	// additional call to SetStatus is required if the Status of the Span should
	// be set to Error, as this method does not change the Span status. If this
	// span is not being recorded or err is nil then this method does nothing.
```

File: trace/trace.go (hunk 1 above)
Comment (dennis-tra) [reply trace/trace.go:222]: I wouldn't know how to associate that :/ 

I guess the association would happen at the level where the transaction is created. We could carry around the span context in the `Txn` struct but then what would you do with the context that gets passed into transaction put/gets?

Comment (Stebalien) [reply trace/trace.go:222]: Yeah, fair enough. I'd just leave it as-is.

File: trace/trace.go (hunk 2 above)
Comment (Stebalien) [reply trace/trace.go:42]: Well, that API is bonkers. But ok.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**